src/ControlSystem/TimescaleTuner.cpp

```diff
@@ -0,0 +1,114 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/TimescaleTuner.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <ostream>
+
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/Gsl.hpp"
+
+TimescaleTuner::TimescaleTuner(DataVector initial_timescale,
+                               double max_timescale, double min_timescale,
+                               double decrease_timescale_threshold,
+                               double increase_timescale_threshold,
+                               double increase_factor,
+                               double decrease_factor) noexcept
+    : timescale_{std::move(initial_timescale)},
+      max_timescale_{max_timescale},
+      min_timescale_{min_timescale},
+      decrease_timescale_threshold_{decrease_timescale_threshold},
+      increase_timescale_threshold_{increase_timescale_threshold},
+      increase_factor_{increase_factor},
+      decrease_factor_{decrease_factor} {
+  for (auto t_scale : timescale_) {
+    if (t_scale <= 0.0) {
+      ERROR("Initial timescale must be > 0");
+    }
+  }
+
+  if (decrease_factor_ > 1.0 or decrease_factor <= 0.0) {
+    ERROR("The specified decrease_factor "
+          << decrease_factor_ << " must satisfy 0 < decrease_factor <= 1");
+  }
+  if (increase_factor_ < 1.0) {
+    ERROR("The specified increase factor " << increase_factor_
+                                           << " must be >= 1.0");
+  }
+  if (min_timescale_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (max_timescale_ <= min_timescale_) {
+    ERROR("The maximum timescale "
+          << max_timescale_
+          << " must be > than the specified minimum timescale "
+          << min_timescale_);
+  }
+  if (increase_timescale_threshold_ <= 0.0) {
+    ERROR("The specified minimum timescale " << min_timescale_
+                                             << " must be > 0");
+  }
+  if (decrease_timescale_threshold_ <= increase_timescale_threshold_) {
+    ERROR("The decrease-timescale threshold "
+          << decrease_timescale_threshold_
+          << " must be > than the specified increase-timescale threshold "
+          << increase_timescale_threshold_);
+  }
+}
+
+void TimescaleTuner::set_timescale_if_in_allowable_range(
+    double suggested_timescale) noexcept {
```

Comment:
const the consts

Reply:
add `cpp17:clamp` see http://en.cppreference.com/w/cpp/algorithm/clamp  into Utilities/ConstantExpressions.hpp